### ISSUE TITLE
Improvement: Merge Custom Highcharts Settings

### ISSUE THREAD

**Issue by buffaloseven** [methnen/m-chart#84]
Per-chart customization could be achieved easier than the theming application if there was a textarea that could be utilized on a per-chart basis to enter a custom JSON highcharts settings object that would then be merged (e.g. via `$.extend()`) into the main chart settings.

**Comment by methnen** [comment]
There's a filter hook that lets you merge any custom settings into the settings being used for each chart. That's essentially what the themes do.

I went the "theme" route basically because I'm leaning towards never directly exposing the guts of Highcharts to the regular user. This was a conscious choice and one I feel somewhat strongly about. The very first version of this plugin was designed for journalists with sometimes little to know experience with coding of any type. Fields in the interface that take direct code input are pretty confusing for that type of person.

That said it would be pretty straightforward to build a plugin that adds a meta box that does exactly what you describe. If you wanted it to take JSON input you could then output the JSON into the chart on the rendering end via an action hook:

https://github.com/methnen/m-chart/wiki/Action-and-filter-hooks#m_chart_post_render_javascript

There's a variety of action and filter hooks that I added with that sort of thing in mind.

If you were interested in making an attempt at it I'd be happy to lend a hand.

**Comment by buffaloseven** [comment]
Hmm, I can see (and agree with) your rationale. The only thing with theming is that, at least to my understanding, the point of a theme would be to create recyclable settings that are utilized multiple times, which would be antithetical, to a point, with per-graph customizations.

Separating the functionality to a plugin for your plugin would be a pleasant compromise, since then you'd have to have an idea of what you wanted to do in order to even bother getting it.

I'll gladly attempt it (I have initiative at the moment since I'm in the middle of producing a lot of graphs and don't necessarily want to create a ton of themes). I have never written a plugin _for a plugin_ before, though; any pointers where to start for learning the general structure?

**Comment by methnen** [comment]
Certainly,

There's basically 3 things you need.

A meta box with your controls, a method to save the input from the meta box and save_post and a method to hook into the appropriate place for you to then use the JSON for a given chart.

The charts are just a custom post type.

So your meta box would just attach to the m-chart post type. Something like this:

```php
add_meta_box(
	'm-chart-json',
	esc_html__( 'JSON', 'm-chart' ),
	array( $this, 'json_meta_box' ),
	'm-chart',
	'normal',
	'high'
);
```

This is assuming you're building the plugin inside of it's own class of course.

The json_meta_box method displays the field that accepts JSON input. And possibly a nonce field if you want to be extra careful. Or you could rely on the nonce I'm already creating.

Then you'll attach to the save_post action with another method:

```php
add_action( 'save_post', array( $this, 'save_post' ) );
```

I like to structure my save_post methods like this so I'm checking for the right things first and then do any necessary work. Something like this:

```php
public function save_post( $post_id ) {
	$post = get_post( $post_id );

	// Check that this isn't an autosave
	if ( defined( 'DOING_AUTOSAVE' ) && DOING_AUTOSAVE ) {
		return;
	}

	// Check post type
	if ( ! isset( $post->post_type ) || 'm-chart' != $post->post_type ) {
		return;
	}

	// Don't run on post revisions (almost always happens just before the real post is saved)
	if ( wp_is_post_revision( $post->ID ) ) {
		return;
	}

	// Make sure we've got some actual M Chart related data in the $_POST array
	if ( ! isset( $_POST['m-chart-json'] ) ) {
		return;
	}

	// Check the nonce (this would check the nonce I already include in the form)
	if ( ! wp_verify_nonce( $_POST['m-chart' ]['nonce'],  'm-chart-save-post' ) ) {
		return;
	}

	// Check the permissions
	if ( ! current_user_can( 'edit_post', $post->ID ) ) {
		return;
	}

	update_post_meta( $post->ID, 'm-chart-json', $_POST['m-chart-json'] );
}
```

Then you need to output it to the chart. Again hooking to an action hook.

```php
add_action( 'm_chart_post_render_javascript', array( $this, 'm_chart_post_render_javascript' ), 10, 3 );
```

Or something along those lines. Then in the `m_chart_post_render_javascript` method you build you'd check to see if you've got some JSON and output it there in whatever method you need.  Since the hook is specific to the charts already you don't need to check post type or anything there like you do when assigning the meta box and saving the data.

As far as general plugin structure I wouldn't do anything different from normal best practices.

Hopefully I'm not just confusing matters or explaining things you already know.

**Comment by methnen** [comment]
Also I'd be happy to let you bug me via Skype whenever you want.

My Skype handle is: methnen

Only caveat is that I'm supposed to have a baby this coming Monday (well my wife is having the baby) so I might be sort of unavailable and/or sleep deprived at times.

**Comment by buffaloseven** [comment]
Thanks for the help. I hope to get started on it in the coming few days.

That said, I seem to have run into a bug with the theming system.  I've created a theme file that looks like this:

```php
<?php

/**
 * Theme Name: Monthly Temperature Summary
 */

$meanTemp = array(
    "color" => "#777777",
    "lineWidth" => 1,
);

$thirtyYear = array(
    "color" => "#222222",
    "lineWidth" => 2,
);

$averageHigh = array(
    "color" => "#E85B56",
    "lineWidth" => 1,
    "visible" => "false"
);

$averageLow = array(
    "color" => "#2A509A",
    "lineWidth" => 1,
    "visible" => "false"
);

$topThirty = array(
    "color" => "#E85B56",
    "type" => "scatter"
);

$bottomThirty = array(
    "color" => "#E85B56",
    "type" => "scatter"
);

$percentile = array(
    "color" => "#FFC640",
    "lineWidth" => "1",
    "showInLeged" => "false"
);

$seriesArray = array($meanTemp, $thirtyYear, $averageHigh, $averageLow, $topThirty, $bottomThirty, $percentile, $percentile);

return array(
    "series" => $seriesArray,
    "tooltip" => array("shared" => true)
);
?>
```

The gist is that I want to apply custom colours, line widths, and types to a complex graph summarizing climatological data.  However, when I use the theme in m-chart, it seems that instead of merging the settings for the series object, it's replacing it. The resultant `chart_args` object when using the theme looks like:

```json
chart_args: {
    "chart": {
        "type": "line",
        "show_labels": false,
        "renderTo": "m-chart-4585",
        "height": 400
    },
    "title": {
        "text": "Untitled"
    },
    "subtitle": {
        "text": ""
    },
    "legend": {
        "enabled": true
    },
    "credits": {
        "href": "",
        "text": ""
    },
    "exporting": {
        "enabled": false
    },
    "plotOptions": {
        "series": {
            "dataLabels": {
                "enabled": false,
                "format": "{y:,f}"
            },
            "connectNulls": true
        },
        "line": {
            "dataLabels": {
                "enabled": false
            }
        }
    },
    "xAxis": {
        "categories": ["1872", "1873", "1874", "1875", "1876", "1877", "1878", "1879", "1880", "1881", "1882", "1883", "1884", "1885", "1886", "1887", "1888", "1889", "1890", "1891", "1892", "1893", "1894", "1895", "1896", "1897", "1898", "1899", "1900", "1901", "1902", "1903", "1904", "1905", "1906", "1907", "1908", "1909", "1910", "1911", "1912", "1913", "1914", "1915", "1916", "1917", "1918", "1919", "1920", "1921", "1922", "1923", "1924", "1925", "1926", "1927", "1928", "1929", "1930", "1931", "1932", "1933", "1934", "1935", "1936", "1937", "1938", "1939", "1940", "1941", "1942", "1943", "1944", "1945", "1946", "1947", "1948", "1949", "1950", "1951", "1952", "1953", "1954", "1955", "1956", "1957", "1958", "1959", "1960", "1961", "1962", "1963", "1964", "1965", "1966", "1967", "1968", "1969", "1970", "1971", "1972", "1973", "1974", "1975", "1976", "1977", "1978", "1979", "1980", "1981", "1982", "1983", "1984", "1985", "1986", "1987", "1988", "1989", "1990", "1991", "1992", "1993", "1994", "1995", "1996", "1997", "1998", "1999", "2000", "2001", "2002", "2003", "2004", "2005", "2006", "2007", "2008", "2009", "2010", "2011", "2012", "2013", "2014", "2015", "2016", "2017"],
        "title": {
            "text": ""
        }
    },
    "yAxis": {
        "title": {
            "text": ""
        }
    },
    "series": [{
        "color": "#777777",
        "lineWidth": 1
    }, {
        "color": "#222222",
        "lineWidth": 2
    }, {
        "color": "#E85B56",
        "lineWidth": 1,
        "visible": "false"
    }, {
        "color": "#2A509A",
        "lineWidth": 1,
        "visible": "false"
    }, {
        "color": "#E85B56",
        "type": "scatter"
    }, {
        "color": "#E85B56",
        "type": "scatter"
    }, {
        "color": "#FFC640",
        "lineWidth": "1",
        "showInLeged": "false"
    }, {
        "color": "#FFC640",
        "lineWidth": "1",
        "showInLeged": "false"
    }],
    "tooltip": {
        "shared": true
    }
}
```

Each entry in the series object should have a default `name` and `data` entry, which shows up correctly with the default themes.  When using mine, the entire series array seems to be overwritten. Any ideas?

**Comment by methnen** [comment]
Yeah, that's a side effect of using array_merge. It's not terribly smart, but it's fast.

Let me think about it a bit and see if I can come up with a solution for your issue.

**Comment by buffaloseven** [comment]
I'm curious; does `$.extend()` already handle this correctly? I can't try until later tonight. If it did, you could simply set a JS theme variable `theme_options = null` and set it if an object is defined in the theme. Then just run the `$.extend()` function if `theme_options !== null`.

Just an idea; if the solution is complex in PHP and a library you have to use for Highcharts handles it, maybe just offload it?

**Comment by methnen** [comment]
If you set deep to true it does:

https://api.jquery.com/jquery.extend/#jQuery-extend-deep-target-object1-objectN

array_merge is really not super smart as I said, and I knew that, but recursively doing it is just more expensive on the server side. In this case it's probably worth it. I'll probably work out a recursive array_merge in the next version.

**Comment by methnen** [comment]
Or now that I've ready your full comment, yeah, that's not a bad idea either.

**Comment by buffaloseven** [comment]
Or [Stack Overflow](http://stackoverflow.com/questions/6975676/php-recursive-merge) to the rescue?

**Comment by buffaloseven** [comment]
So this seems to work. In `class-m-chart-highcharts.php` I added:

```php
// Recursive merge function provided at http://stackoverflow.com/questions/6975676/php-recursive-merge
function recursive_merge(&$a,$b){ //$a will be result. $a will be edited. It's to avoid a lot of copying in recursion
    foreach($b as $child=>$value){
        if(isset($a[$child])){ 
            if(is_array($a[$child]) && is_array($value)){ //merge if they are both arrays
                recursive_merge($a[$child],$value);
            }
            //else ignore, you can add your own logic, i.e when 1 of them is array
        }
        else
            $a[$child]=$value; //add if not exists
    }
    //return $a;
}
```

And replaced
```php
188			$chart_args = array_merge( $chart_args, $theme );
```

with

```php
188			recursive_merge( $chart_args, $theme );
```

It works perfectly (or so it seems).  Should probably do a little more testing, but I wasn't able to find any conflicts so far.

**Comment by buffaloseven** [comment]
I'm successfully merging a theme file with numerous series options with nested associative arrays in them and it seems to be holding up well.  This fairly complicated chart works exactly as expected.
<img width="936" alt="screen shot 2016-12-02 at 9 28 36 pm" src="https://cloud.githubusercontent.com/assets/3456575/20856537/cae326ea-b8d6-11e6-9c03-43dd82827231.png">

It has custom colours, custom line widths, markers, plot types, tooltip formatters, legend visibility, z-index values, and hover states.  No apparent problems or conflicts 👍

**Comment by methnen** [comment]
Awesome. I'll look at the method you ended up using and probably implement something along those lines in the next release. Hopefully sometime this month. I'm in the midst of some other work on the plugin that I'm pretty close to finishing up and then I can move to this next.

**Comment by buffaloseven** [comment]
Sounds great!

As an aside, is there a way to call a chart via Javascript? I'm making some pages where dynamic loading of charts could be handy.

**Comment by methnen** [comment]
There currently isn't.  However, the next version will be able to load charts via iFrames.

That same functionality could be leveraged to dynamically load charts pretty easily I think.

**Comment by buffaloseven** [comment]
I like the sound of that! :)

**Comment by methnen** [comment]
Closing as relevant issue is addressed in 1.6.
